cmd/api-utils.go

```diff
@@ -0,0 +1,107 @@
+/*
+ * Minio Cloud Storage, (C) 2019 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"strings"
+)
+
+func shouldEscape(c byte) bool {
+	if 'A' <= c && c <= 'Z' || 'a' <= c && c <= 'z' || '0' <= c && c <= '9' {
+		return false
+	}
+
+	switch c {
+	case '-', '_', '.', '/', '*':
+		return false
+	}
+	return true
+}
+
+// s3URLEncode is based on Golang's url.QueryEscape() code,
+// while considering some S3 exceptions:
+//	- Avoid encoding '/' and '*'
+//	- Force encoding of '~'
+func s3URLEncode(s string) string {
+	spaceCount, hexCount := 0, 0
+	for i := 0; i < len(s); i++ {
+		c := s[i]
+		if shouldEscape(c) {
+			if c == ' ' {
+				spaceCount++
+			} else {
+				hexCount++
+			}
+		}
+	}
+
+	if spaceCount == 0 && hexCount == 0 {
+		return s
+	}
+
+	var buf [64]byte
+	var t []byte
+
+	required := len(s) + 2*hexCount
+	if required <= len(buf) {
+		t = buf[:required]
+	} else {
+		t = make([]byte, required)
+	}
+
+	if hexCount == 0 {
```

Comment:
is this check required? is the space check not handled in the next for loop?

Reply:
> is this check required? is the space check not handled in the next for loop?

This code is based in Golang's url.QueryEscape(), but I will add some benchmark to check it.

Reply:
> This code is based in Golang's url.QueryEscape(), but I will add some benchmark to check it.

```
18:02 $ go test -bench=BenchmarkS3EncodeURL
goos: linux
goarch: amd64
pkg: github.com/minio/minio/cmd
BenchmarkS3EncodeURL-4           5000000               241 ns/op
BenchmarkS3EncodeURLFast-4       5000000               390 ns/op
PASS
ok      github.com/minio/minio/cmd      109.138s
```

Without that, it is more slower since paths most likely don't contain special characters like '{', '%', etc.. so I think we are good to keep this